### PR TITLE
xhyve: bump version to support Big Sur

### PR DESCRIPTION
**Pull request by catap**
#### Description

This is moving to a few commits ahead mainly to include https://github.com/machyve/xhyve/pull/200 which introduced support of Big Sur

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 11.5.1 20G80 x86_64
Xcode 12.5.1 12E507

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: emulators/xhyve/Portfile
Comment (reneeotten) [inline emulators/xhyve/Portfile:31]: why are these `reinplaces` not necessary anymore?

Comment (catap) [reply emulators/xhyve/Portfile:31]: @reneeotten when I tried to build it locally, I have a warning:
```
--->  Extracting xhyve
Warning: reinplace s|build/xhyve|/opt/local/bin/xhyve| didn't change anything in /opt/local/var/macports/build/_Users_catap_ports_emulators_xhyve/xhyve/work/xhyve-83516a009c692ea5d2993d1071e68d05d359b11e/xhyverun-tinycorelinux.sh
Warning: reinplace s|build/xhyve|/opt/local/bin/xhyve| didn't change anything in /opt/local/var/macports/build/_Users_catap_ports_emulators_xhyve/xhyve/work/xhyve-83516a009c692ea5d2993d1071e68d05d359b11e/xhyverun-freebsd.sh
Warning: reinplace s|test/|/opt/local/share/xhyve/test/| didn't change anything in /opt/local/var/macports/build/_Users_catap_ports_emulators_xhyve/xhyve/work/xhyve-83516a009c692ea5d2993d1071e68d05d359b11e/xhyverun-windows.sh
Warning: reinplace s|build/xhyve|/opt/local/bin/xhyve| didn't change anything in /opt/local/var/macports/build/_Users_catap_ports_emulators_xhyve/xhyve/work/xhyve-83516a009c692ea5d2993d1071e68d05d359b11e/xhyverun-windows.sh
--->  Configuring xhyve
```

as far as I see it do not needed since https://github.com/machyve/xhyve/commit/1dd9a5165848c7ed56dafc41932c553ea56a12af#diff-a282a5eb55d8da9fb9e95d97048f60dd145519306facc437d60813da23df091e

Comment (reneeotten) [reply emulators/xhyve/Portfile:31]: yes, but there are three of `.sh` scripts and therefore a total of 6 reinplaces that are executed - apparently, four of these don't do anything, but two of them do. Of course you can figure out which reinplaces you need to keep.

Comment (catap) [reply emulators/xhyve/Portfile:31]: @reneeotten god catch! Reverting back.

**Comment by catap** [comment]
About CI: it fails with very wired issue:
```
2021-08-17T21:01:38.6978470Z While building module 'Hypervisor' imported from /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_emulators_xhyve/xhyve/work/machyve-xhyve-83516a0/src/vmm/intel/vmx_msr.c:34:
2021-08-17T21:01:38.6980060Z In file included from <module-includes>:1:
2021-08-17T21:01:38.6981230Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:67:35: error: expected function body after function declarator
2021-08-17T21:01:38.6982310Z         uint32_t field, uint64_t *value) __HV_10_15;
2021-08-17T21:01:38.6982700Z                                          ^
2021-08-17T21:01:38.6984050Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:82:34: error: expected function body after function declarator
2021-08-17T21:01:38.6985190Z         uint32_t field, uint64_t value) __HV_10_15;
2021-08-17T21:01:38.6985570Z                                         ^
2021-08-17T21:01:38.6986580Z /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:109:43: error: expected function body after function declarator
2021-08-17T21:01:38.6987860Z         uint32_t field, hv_shadow_flags_t flags) __HV_10_15;
2021-08-17T21:01:38.6988270Z                                                  ^
2021-08-17T21:01:38.6988640Z 3 errors generated.
2021-08-17T21:01:38.6990270Z /opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_emulators_xhyve/xhyve/work/machyve-xhyve-83516a0/src/vmm/intel/vmx_msr.c:34:10: fatal error: could not build module 'Hypervisor'
2021-08-17T21:01:38.6991390Z #include <Hypervisor/hv.h>
2021-08-17T21:01:38.6991760Z  ~~~~~~~~^
2021-08-17T21:01:38.7039370Z 4 errors generated.
```

This macros is defined like this:
```
#define __HV_10_10 __OSX_AVAILABLE_STARTING(__MAC_10_10, __IPHONE_NA)
#define __HV_10_15 __OSX_AVAILABLE_STARTING(__MAC_10_15, __IPHONE_NA)
```
and I feel that a root cause of this issue is github's virtualisation.

File: emulators/xhyve/Portfile
Comment (reneeotten) [inline emulators/xhyve/Portfile:30]: please align the patchfiles in the same was as the rest of the Portfile and remove the extraneous newline

**Comment by reneeotten** [comment]
I don't know anything about this port nor really understand the error message you show... so I'm afraid that I cannot be if much help for this.

**Comment by catap** [comment]
@reneeotten seems that it is "feature" of clang 12 on macOS 10.15.

I have create he simpler possible reproducer for this error:
```
catap@catapsmCatalina /tmp % cat test.c 
#include <Hypervisor/hv.h>


catap@catapsmCatalina /tmp % clang -fmodules test.c
While building module 'Hypervisor' imported from test.c:1:
In file included from <module-includes>:1:
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:67:35: error: expected function body after function
      declarator
        uint32_t field, uint64_t *value) __HV_10_15;
                                         ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:82:34: error: expected function body after function
      declarator
        uint32_t field, uint64_t value) __HV_10_15;
                                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Hypervisor.framework/Headers/hv_vmx.h:109:43: error: expected function body after function
      declarator
        uint32_t field, hv_shadow_flags_t flags) __HV_10_15;
                                                 ^
test.c:1:10: fatal error: could not build module 'Hypervisor'
#include <Hypervisor/hv.h>
 ~~~~~~~~^
4 errors generated.
catap@catapsmCatalina /tmp % clang --version
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
catap@catapsmCatalina /tmp % 
```

If using clang-11 it works as expected. clang-12 introduced the full modules support as `-fmodules` and an old one is renamed to `-fmodules-ts`.

Anyway, I have no idea how to replace `-fmodules` to `-fmodules-ts` on xcode project based build such we have here.

Maybe you can addvice or tag someone who can?

Thanks!

**Comment by catap** [comment]
Wired. I can build it on:
```
macOS 10.15.7 19H15 x86_64
Xcode 12.4 12D4e
```

**Review by reneeotten — COMMENTED** [review]
@catap as I said earlier I'm sorry but I don't have the expertise to help you out on this. You could ask help on the mailinglist - there are likely people out there who would understand what's going on.

**Comment by catap** [comment]
@reneeotten can it be merged without passing some old system? :)

**Comment by reneeotten** [comment]
> @reneeotten can it be merged without passing some old system? :)

I think the "policy" is to support the latest three macOS releases, anything older isn't hard requirement but we do attempt to support them as much as possible. This does not build on the CI for 10.15 so that's likely something that needs to be fixed before we'd consider merging the PR.

**Comment by catap** [comment]
@reneeotten and if the old port doesn't build on modern CI system? The issue is Xcode 12.4? I'd like to revert all my changes to just bump a revision number to proof that ;)

**Comment by catap** [comment]
@reneeotten you may see that CI fails on the same issue on current code. I have no idea how to fix it, because the bug is inside Xcode 12.4 :(

**Comment by reneeotten** [comment]
@kencu @cjones051073 has any of you seen this issue before and/or is able to take a look? I'm afraid this is above my pay-grade and I cannot really help here anymore...

**Comment by kencu** [comment]
A mismatch between headers that are pulled in would seem to be likely at play. Exactly what mismatch I am not yet clear on, but   ```__HV_10_15``` is supposed to be defined but is not being defined.

A similar issue with modules occurred with the ncurses port. As I recall how it worked (18 months ago) when the modules are being used the build system did not follow the expected order of including headers, and in ncurses, the wrong (system) header was pulled in instead of the one we wanted to use (in /opt/local/include).

https://trac.macports.org/ticket/59992

the temporary solution, at least, was to add a module.map for ncurses that was installed in /opt/local/include that would service ncurses.

https://github.com/macports/macports-ports/commit/d2c8da96efe60cc3a72c03c92aa2d4af8eb280d4#diff-4f1aa1994e32ae4d841cca3766dec0d31cd5c4cd6aeeb07dd114bc9b1b49a4cb

It was obvious that this was going to be the tip of the iceberg for this issue, and ncurses was only going to be the first of many such modules issues. How that was ever going to be coherently sorted out for the general case was unknown.

**Comment by catap** [comment]
@kencu I guess I need some help or advice to make required module.map.

`__HV_10_15` is defined inside `hv.h` but it also used inside `hv_vmx.h`.

I really do not understand that's going on.

**Comment by kencu** [comment]
There are not many who have a deep understanding of modules yet, and to make it worse, I think the implementation is still in flux. I certainly would not want to appear to be any kind of expert on the subject, assuming that module header confusion is what is really going on here.

**Comment by catap** [comment]
@kencu right now this port is broken for both: 10.15 and 11 macOS. My changes fixes it for 11 that is good enough for me.

I think that adding some hacks to overstep issues with modules is good idea, but it seems like a separated PR for me.

Also, with comparing with a workaround which was made for ncurses I see here more ticky case:
 1. It isn't simple header which need to be wrapped, it is OS framework;
 2. build system is using xcode and a hack should be incorporated to xcode project or at least I think so.

Here I suggest to remove module map changes because it is useless for now and move forward with fixes port for at least Big Sur that was original goal of this PR.

**Comment by kencu** [comment]
don't pound on me for trying to be helpful.

I've never used this port, and have no opinions about whether this PR is a step forward or back.

signing off.

**Comment by catap** [comment]
@reneeotten I guess I figured it out and here a question to which you seems the right person to ask :)

How can I enforce to use xcode sdk and not CommandLineTools?

macOS 11 bild uses xcode:
```
2021-09-05T00:27:26.8873120Z LIBRARY_PATH='/opt/local/lib'
2021-09-05T00:27:26.8874050Z MACOSX_DEPLOYMENT_TARGET='11.0'
2021-09-05T00:27:26.8875380Z SDKROOT='/Applications/Xcode_12.5.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk'
```
when macOS 10.15 is using CommandLineTools:
```
2021-09-05T00:27:19.4900720Z LIBRARY_PATH='/opt/local/lib'
2021-09-05T00:27:19.4901600Z MACOSX_DEPLOYMENT_TARGET='10.15'
2021-09-05T00:27:19.4902670Z SDKROOT='/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk'
```

**Comment by reneeotten** [comment]
@catap if the port really needs a full XCode installation you can specify `use_xcode yes`. I am not sure though it that will do whatever you'd like it to do, but it's worth trying.

**Comment by catap** [comment]
@reneeotten I expect that port with xcode group do not need it. Anyway, try it cost nothing... and it doesn't work :(

**Comment by catap** [comment]
@reneeotten and I figured it out!

**Comment by catap** [comment]
@reneeotten hello here!

I have two open questions about this PR / port:
 - can it be merged? :)
 - because I've started to claim maintainership for some ports I can also do for this one which I'm using. Anyway, I haven't see any explanation how to do it when a port has maintainer.

**Comment by catap** [comment]
@pmetzger can I ask your thought about this one?

Current version can't be build on both: 10.15 and 11 :) This one fixed it.

**Comment by kencu** [comment]
apologies this is taking a bit of time. 

I presume you already have it locally installed, so you are using it now. So you're all set. This PR is for the general population out there, and thanks for that contribution. 

There seem to be a few things that will need verifying that the fixes are the ones needed long-term to support general users, and it will take a bit of time for that to happen. 

Please don't be frustrated while it is looked over.

**Comment by pmetzger** [comment]
It looks like we're now at the CI passing and Jeremy has clearly abdicated management of this port (it might be reasonable to open a "port abandoned" ticket fwiw.) I'd suggest that if no one has anything significant to say that we merge this in the next few days.

**Comment by pmetzger** [comment]
@catap btw, you should probably apply for commit privileges. You're certainly working on MacPorts more than enough for that.

**Comment by catap** [comment]
@pmetzger I've already started the process about take care of this port: https://trac.macports.org/ticket/63495

and regarding commit privileges, I'd like to agree with you.

File: emulators/xhyve/files/add-hypervisor-elements.diff
Comment (jeremyhu) [inline emulators/xhyve/files/add-hypervisor-elements.diff:1]: This has been done upstream.  Can you drop the patch and pick up newer upstream?

File: emulators/xhyve/files/disable-VMX-big-sur.diff
Comment (jeremyhu) [inline emulators/xhyve/files/disable-VMX-big-sur.diff:1]: This has also been merged upstream, so just update to a newer upstream commit.

**Review by jeremyhu — CHANGES_REQUESTED** [review]

**Comment by catap** [comment]
@jeremyhu, as you make clear in https://trac.macports.org/ticket/63495 I've made myself also maintaier of this port.

Thus, the new version contains small fixes for clang `clang-1300.0.29.3` which is shiped with 11.6. A patch was backported to upstream: https://github.com/machyve/xhyve/pull/219

@jeremyhu let me know if you plan to merge the patch to upstream to make one more bump :)

File: emulators/xhyve/Portfile
Comment (pmetzger) [inline emulators/xhyve/Portfile:13]: @catap I suggest adding "openmaintainer" too.

**Review by pmetzger — CHANGES_REQUESTED** [review]
One small comment.

File: emulators/xhyve/Portfile
Comment (catap) [reply emulators/xhyve/Portfile]: @pmetzger let make things easy for everyone, yeah.

**Review by pmetzger — APPROVED** [review]
LGTM.

File: emulators/xhyve/files/cast-int.patch
Comment (jeremyhu) [inline emulators/xhyve/files/cast-int.patch:1]: Could you upstream this as well?  I didn't see it when quickly looking through the pending pull requests upstream last night (I could have missed it).

**Review by jeremyhu — APPROVED** [review]

File: emulators/xhyve/files/cast-int.patch
Comment (catap) [reply emulators/xhyve/files/cast-int.patch:1]: @jeremyhu I'm usually send to upstream all patches that I'm making. For example this one is here: https://github.com/machyve/xhyve/pull/219

Comment (catap) [reply emulators/xhyve/files/cast-int.patch:1]: BTW, it is quite new. Because without this patch I can compile it on Big Sur until 11.6.

**Comment by pmetzger** [comment]
Thanks, @catap, and do continue to monitor for updates that might do things like incorporating your patches.

**Pull request merged**
